### PR TITLE
JPERF-487: Enabled custom logging configuration

### PR DESCRIPTION
**Pull request by lipinskipawel**
According to https://ecosystem.atlassian.net/browse/JPERF-487

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -84,6 +84,7 @@ dependencies {
     implementation("org.apache.commons:commons-csv:1.3")
     implementation("commons-cli:commons-cli:1.4")
 
+    api(log4j("core"))
```
Comment (sebapawlak) [inline build.gradle.kts]: Since this is becoming a part of API I think we need a `CHANGELOG` entry for this change too.
> - Add `log4j-core` to the API.

File: CHANGELOG.md [hunk 2]
```diff
@@ -23,6 +23,11 @@ Dropping a requirement of a major version of a dependency is a new contract.
 ## [Unreleased]
 [Unreleased]: https://github.com/atlassian/virtual-users/compare/release-3.6.6...master
 
+### Added
+- Add custom logging configuration [JPERF-487].
```
Comment (sebapawlak) [inline CHANGELOG.md]: Could you please stick to the existing convention?
Example:
>Daemonize WebDriver closing threads. Fix JPERF-470.

So it should be something like:

> Add custom logging configuration. Resolve [JPERF-487].

File: src/main/kotlin/com/atlassian/performance/tools/virtualusers/api/EntryPoint.kt [hunk 3]
```diff
@@ -28,12 +33,22 @@ class Application {
...
-        if (options.help) {
-            options.printHelp()
-        } else {
-            LoadTest(options, RestUserGenerator()).run()
-        }
+        val options: VirtualUserOptions = VirtualUserOptions.Parser().parse(args)
+
+        setLoggingConfiguration(options)
+
+        if (options.help) options.printHelp()
```
Comment (sebapawlak) [inline src/main/kotlin/com/atlassian/performance/tools/virtualusers/api/EntryPoint.kt]: You have removed the `@Suppress("DEPRECATION")` which means there is one more warning here. Please revert the change unless there is some specific reason. If so please explain.

File: src/main/kotlin/com/atlassian/performance/tools/virtualusers/api/EntryPoint.kt [hunk 4]
```diff
@@ -28,12 +33,22 @@ class Application {
...
\ No newline at end of file
+
+    private fun setLoggingConfiguration(options: VirtualUserOptions) {
+        val logging = options.behavior.logging
+
+        val logConfig = logging.getConstructor().newInstance() as AbstractConfiguration
+        val logConfigurationFactory = LogConfigurationFactory(logConfig)
+        setConfigurationFactory(logConfigurationFactory)
+        val ctx = LogManager
+            .getContext(false) as org.apache.logging.log4j.core.LoggerContext
```
Comment (sebapawlak) [inline src/main/kotlin/com/atlassian/performance/tools/virtualusers/api/EntryPoint.kt]: Can we import `org.apache.logging.log4j.core.LoggerContext`?

File: src/main/kotlin/com/atlassian/performance/tools/virtualusers/api/EntryPoint.kt (hunk 3 above)
Comment (sebapawlak) [reply src/main/kotlin/com/atlassian/performance/tools/virtualusers/api/EntryPoint.kt]: Also please revert back the `if-else` with braces. We want to avoid problems like this one: https://stackoverflow.com/questions/8020228/is-it-ok-if-i-omit-curly-braces-in-java

**Comment by sebapawlak** [comment]
WRT the commit message: I know it is a nit-pick but we tend to use imperative mood as per https://chris.beams.io/posts/git-commit/#imperative. So ideally the commit message should say:`Enable...` not `Enabled...`.

File: src/main/kotlin/com/atlassian/performance/tools/virtualusers/api/EntryPoint.kt [hunk 5]
```diff
@@ -28,12 +33,22 @@ class Application {
...
-}
\ No newline at end of file
+
+    private fun setLoggingConfiguration(options: VirtualUserOptions) {
+        val logging = options.behavior.logging
+
+        val logConfig = logging.getConstructor().newInstance() as AbstractConfiguration
+        val logConfigurationFactory = LogConfigurationFactory(logConfig)
+        setConfigurationFactory(logConfigurationFactory)
+        val ctx = LogManager
```
Comment (sebapawlak) [inline src/main/kotlin/com/atlassian/performance/tools/virtualusers/api/EntryPoint.kt]: Let's name it `context`. `ctx` is meh...

File: CHANGELOG.md (hunk 2 above)
Comment (lipinskipawel) [reply CHANGELOG.md]: Sure

File: src/main/kotlin/com/atlassian/performance/tools/virtualusers/api/EntryPoint.kt (hunk 4 above)
Comment (lipinskipawel) [reply src/main/kotlin/com/atlassian/performance/tools/virtualusers/api/EntryPoint.kt]: Sure

**Comment by sebapawlak** [comment]
@lipinskipawel this is some solid and good job done here. The comments I have left are more of a nit-picks. I don't see any serious issues. Please address them and wait for feedback from one more person and we should be good to go!

File: src/main/kotlin/com/atlassian/performance/tools/virtualusers/api/EntryPoint.kt (hunk 3 above)
Comment (lipinskipawel) [reply src/main/kotlin/com/atlassian/performance/tools/virtualusers/api/EntryPoint.kt]: The main reason why I changed it was simply to reduce lines of code.
Before 6 now 2 it is more readable, but I don't have any strong opinion why I shouldn't revert those lines, so Sure I will revert.

File: build.gradle.kts (hunk 1 above)
Comment (lipinskipawel) [reply build.gradle.kts]: Sure

File: src/main/kotlin/com/atlassian/performance/tools/virtualusers/api/EntryPoint.kt (hunk 3 above)
Comment (sebapawlak) [reply src/main/kotlin/com/atlassian/performance/tools/virtualusers/api/EntryPoint.kt]: Less lines is not always more readable.

File: src/main/kotlin/com/atlassian/performance/tools/virtualusers/api/EntryPoint.kt [hunk 6]
```diff
@@ -16,8 +20,10 @@ fun main(args: Array<String>) {
 }
 
 class Application {
+    private val defaultLogConfigFactory = LogConfigurationFactory(LogConfiguration())
+
     fun tryRunning(args: Array<String>) {
-        setConfigurationFactory(LogConfigurationFactory())
+        setConfigurationFactory(defaultLogConfigFactory)
```
Comment (dagguh) [inline src/main/kotlin/com/atlassian/performance/tools/virtualusers/api/EntryPoint.kt]: I was worried about the "chicken or the egg" problem, but using a default pre-parsing solves it nicely
Well done 🙂

File: src/main/kotlin/com/atlassian/performance/tools/virtualusers/api/VirtualUserOptions.kt [hunk 7]
```diff
@@ -405,9 +417,16 @@ class VirtualUserOptions(
...
+            } else GoogleChrome::class.java
+        }
+
+        private fun getLogging(commandLine: CommandLine): Class<out AbstractConfiguration> {
+            return if (commandLine.hasOption(loggingParameter)) {
+                val logging = commandLine.getOptionValue(loggingParameter)
+                val loggingClass = Class.forName(logging)
+                val loggingConstructor = loggingClass.getConstructor()
+                (loggingConstructor.newInstance() as AbstractConfiguration)::class.java
+            } else LogConfiguration::class.java
```
Comment (dagguh) [inline src/main/kotlin/com/atlassian/performance/tools/virtualusers/api/VirtualUserOptions.kt]: If you dislike an uneven `if-else` structure, we can go for:

### Good old extraction
```kotlin
private fun getLogging(commandLine: CommandLine): Class<out AbstractConfiguration> {
    return if (commandLine.hasOption(loggingParameter)) {
        parseLoggingClass(commandLine)
    } else {
        LogConfiguration::class.java
    }
}

private fun parseLoggingClass(commandLine: CommandLine): Class<out AbstractConfiguration> {
    val logging = commandLine.getOptionValue(loggingParameter)
    val loggingClass = Class.forName(logging)
    val loggingConstructor = loggingClass.getConstructor()
    return (loggingConstructor.newInstance() as AbstractConfiguration)::class.java
}
```

### Single expression
```kotlin
private fun getLogging(
    commandLine: CommandLine
): Class<out Configuration> = commandLine
    .getOptionValue(loggingParameter)
    ?.let { Class.forName(it) }
    ?.let { it.getConstructor() }
    ?.let { it.newInstance() as Configuration }
    ?.let { it::class.java }
    ?: LogConfiguration::class.java
```

### Accept it
Or we can just live with a slightly uneven if-else blocks. 8 lines is not much, it's basically a short snippet.

Comment (lipinskipawel) [reply src/main/kotlin/com/atlassian/performance/tools/virtualusers/api/VirtualUserOptions.kt]: Maybe I misunderstand something but does it mean that methods like getScenario and getBrowser will remain as not refactored (good old extraction)?
I like the second approach (Single expression) it looks more like functional code which I found that type of style programming across the repository.

Comment (dagguh) [reply src/main/kotlin/com/atlassian/performance/tools/virtualusers/api/VirtualUserOptions.kt]: You can apply the refactor to nearby code for consistency
